### PR TITLE
Add support for Shiny Express in Quarto Dashboards, take 2

### PR DESCRIPTION
**Pull request by wch**
This PR supersedes #1212. It adds support for Shiny Express apps in Quarto Dashboards. Prior to this PR, it would throw an error if a Quarto document had an import from `shiny.express`.

For example:

````qmd
---
title: "Palmer Penguins"
format: dashboard
server: shiny
---

```{python}
#| context: setup
import seaborn as sns
from shiny.express import render, ui, input
penguins = sns.load_dataset("penguins")
```

# {.sidebar}

```{python}
species = list(penguins["species"].value_counts().index)
ui.input_checkbox_group(
    "species", "Species:",
    species, selected = species
)
```

# Plots

```{python}
@render.plot
def depth():
    data = penguins[penguins["species"].isin(input.species())]
    return sns.displot(
        data, x = "bill_depth_mm",
        hue = "species", kind = "kde",
        fill = True
    )
```
````

This is what it looks like:

image

The generated `app.py` looks like this.

```py
# This file generated by Quarto; do not edit by hand.
# shiny_mode: core

from __future__ import annotations

from pathlib import Path
from shiny import App, Inputs, Outputs, Session, ui

import seaborn as sns
from shiny import reactive
from shiny.express import render, ui
penguins = sns.load_dataset("penguins")

# ========================================================================




def server(input: Inputs, output: Outputs, session: Session) -> None:
    # from shiny.express import render, ui
    species = list(penguins["species"].value_counts().index)
    ui.input_checkbox_group(
        "species", "Species:",
        species, selected = species
    )

    # ========================================================================

    @render.plot
    def depth():
        data = penguins[penguins["species"].isin(input.species())]
        return sns.displot(
            data, x = "bill_depth_mm",
            hue = "species", kind = "kde",
            fill = True
        )

    # ========================================================================



    return None


_static_assets = ["simple_files"]
_static_assets = {"/" + sa: Path(__file__).parent / sa for sa in _static_assets}

app = App(
    Path(__file__).parent / "simple.html",
    server,
    static_assets=_static_assets,
)
```

A few things to note about this:

- It uses the UI was generated by Quarto (in the form of a static .html file).
- The presence of a magic comment `# shiny_mode: express` or `# shiny_mode: core` will force the app to be run as an Express or Core app. This comment must be in the first 1000 characters of the app file.
- The `is_express_app()` function has changed. There are other places where we use copies of this code, (Shiny Server and rsconnect-python), and the function will need to be updated in those places.

### CODE REVIEW

**Comment by wch** [comment]
- Use a magic comment to force app to be treated as Core or Express app.
- Warn if `page_opts()` or `app_opts()` are called?

File: shiny/express/_is_express.py [hunk 1]
```diff
@@ -44,6 +46,14 @@ def is_express_app(app: str, app_dir: str | None) -> bool:
         # Read the file, parse it, and look for any imports of shiny.express.
         with open(app_path) as f:
```
Comment (jcheng5) [inline shiny/express/_is_express.py:47]: This needs to be UTF-8 but you'll do that in the other PR I assume

File: shiny/express/_is_express.py [hunk 2]
```diff
@@ -78,3 +88,24 @@ def visit_Module(self, node: ast.Module) -> None:
...
+    core".
+
+    Returns
+    -------
+    :
+        `True` if Shiny Express comment is found, `False` if Shiny Core comment is
+        found, and `None` if no magic comment is found.
+    """
+
+    for line in content[:1000].splitlines():
```
Comment (jcheng5) [inline shiny/express/_is_express.py:105]: If you use the `re.MULTILINE` flag you don't need to split (although you'd need to change `\s` to `[ \t]`

File: shiny/express/_run.py [hunk 3]
```diff
@@ -176,6 +183,9 @@ def reset_top_level_recall_context_manager() -> None:
 
 
 def get_top_level_recall_context_manager() -> RecallContextManager[Tag]:
+    if _top_level_recall_context_manager is None:
+        raise RuntimeError("No top-level recall context manager has been set.")
```
Comment (jcheng5) [inline shiny/express/_run.py:187]: Is there a reasonable way you expect users could hit this condition? (And if so, would this error message help them fix it?)

File: shiny/express/ui/_page.py [hunk 4]
```diff
@@ -58,7 +58,14 @@ def page_opts(
...
         heuristics for choosing page functions.
     """
-    cm = get_top_level_recall_context_manager()
+    try:
+        cm = get_top_level_recall_context_manager()
+    except RuntimeError:
+        # We can get here if a Shiny Core app, or if we're in the UI rendering phase of
+        # a Quarto-Shiny dashboard.
+        raise RuntimeError(
+            "express.ui.page_opts() can only be used inside of a standalone Shiny Express app."
```
Comment (jcheng5) [inline shiny/express/ui/_page.py:67]: Oh, I see.

**Review by jcheng5 — APPROVED** [review]

**Pull request merged**